File: 15/Solution.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new LinkedList<>();
        Arrays.sort(nums);

        int length = nums.length;

        for(int i = 0; i < length - 2; i++) {
            if(i > 0 && nums[i] == nums[i - 1])
                continue;
            int l = i + 1;
            int r = length - 1;

            while(l < r) {
                int total = nums[i] + nums[l] + nums[r];
                if(total < 0) {
                    l += 1;
                } else if(total > 0) {
                    r -= 1;
                } else {
                    res.add(Arrays.asList(nums[i], nums[l], nums[r]));
                    while(l < r && nums[l] == nums[l + 1]) {
                        l += 1;
                    }
                    while(l < r && nums[r] == nums[r - 1]) {
                        r -= 1;
                    }

                    l += 1;
                    r -= 1;
                }
            }
        }
        return res;
    }
}

File: 15/Solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:

        res = []
        nums.sort()

        length = len(nums)

        for i in range(length - 2):
            if > 0 and nums[i] == nums[i - 1]:
                continue
            l = i + 1
            r = length - 2

            while l < r:
                total = nums[i] + nums[l] + nums[r]
                if total < 0:
                    l += 1
                elif total > 0:
                    r -= 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    while l < r and nums[r] == nums[r - 1]:
                        r -= 1
                l += 1
                r -= 1
        return res

